client/allocrunner/consulsock_hook.go

```diff
@@ -0,0 +1,319 @@
+		return nil
+	case <-time.After(3 * time.Second):
+		return fmt.Errorf("timed out waiting for proxy to exit")
+	}
+}
+
+// Proxy between a listener and dest
+func proxy(ctx context.Context, logger hclog.Logger, dest string, l net.Listener) {
+	// Wait for all connections to be done before exiting to prevent
+	// goroutine leaks.
+	wg := sync.WaitGroup{}
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		// Must cancel context and close listener before waiting
+		cancel()
+		l.Close()
+		wg.Wait()
+	}()
+
+	// Close Accept() when context is cancelled
+	go func() {
+		<-ctx.Done()
+		l.Close()
+	}()
+
+	for ctx.Err() == nil {
+		conn, err := l.Accept()
+		if err != nil {
+			if ctx.Err() != nil {
+				// Accept errors during shutdown are to be expected
+				return
+			}
+			logger.Error("error in grpc proxy; shutting down proxy", "error", err, "dest", dest)
+			return
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			proxyConn(ctx, logger, dest, conn)
+		}()
+	}
+}
+
+// proxyConn proxies between an existing net.Conn and a destination address. If
+// the destAddr starts with "unix://" it is treated as a path to a unix socket.
+// Otherwise it is treated as a host for a TCP connection.
+//
+// When the context is cancelled proxyConn blocks until all goroutines shutdown
+// to prevent leaks.
+func proxyConn(ctx context.Context, logger hclog.Logger, destAddr string, conn net.Conn) {
+	// Close the connection when we're done with it.
+	defer conn.Close()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Detect unix sockets
+	network := "tcp"
+	const unixPrefix = "unix://"
+	if strings.HasPrefix(destAddr, unixPrefix) {
+		network = "unix"
+		destAddr = destAddr[len(unixPrefix):]
+	}
+
+	dialer := &net.Dialer{}
+	dest, err := dialer.DialContext(ctx, network, destAddr)
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		logger.Trace("proxy exiting gracefully", "error", err, "dest", destAddr,
+			"src_local", conn.LocalAddr(), "src_remote", conn.RemoteAddr())
+		return
+	}
+	if err != nil {
+		logger.Error("error connecting to grpc", "error", err, "dest", destAddr)
+		return
+	}
+
+	// Wait for goroutines to exit before exiting to prevent leaking.
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+
+	// socket -> gRPC
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		n, err := io.Copy(dest, conn)
+		if ctx.Err() == nil && err != nil {
+			logger.Error("error proxying to Consul", "error", err, "dest", destAddr,
+				"src_local", conn.LocalAddr(), "src_remote", conn.RemoteAddr(),
+				"bytes", n,
+			)
+			return
+		}
+		logger.Trace("proxy to Consul complete",
+			"src_local", conn.LocalAddr(), "src_remote", conn.RemoteAddr(),
+			"bytes", n,
+		)
+	}()
+
+	// gRPC -> socket
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		n, err := io.Copy(conn, dest)
+		if ctx.Err() == nil && err != nil {
+			logger.Trace("error proxying from Consul", "error", err, "dest", destAddr,
```

Comment:
> Is having Trace here rather than Error as above intentional to avoid duplicate log messages?

Yes. Although now you have me wondering if it's possible for this to happen:

1. Everything is running fine (proxy + Envoy + Consul)
2. Copy(envoy <- consul) errors
3. Connections are closed
4. Copy(consul <- envoy) gets an EOF and exits successfully (!!)

In this case the proxy could be dead and Envoy may never know it. Everything would suspect it's still healthy and it would be incredibly difficult to diagnose.

I'm unsure what errors are possible for net.Conns to return from Read and Write when they're closed. AFAIK it's implementation (eg TCPConn vs unix conn, Windows vs Linux) specific, so I think I should assume the above _is_ possible. :thinking:

Reply:
Yeah I'm not sure if a TCP timeout returns EOF here?

The other way I'd expect an error is on a signal interrupt, but at that point I'd expect both sides to receive it and the Nomad process itself is being interrupted anyways.

Reply:
For now I'm making it WARN in both cases. I spent some time working on an alternate approach but wasn't happy with the immediate results and how hard they were to test.

I want to get this merged and see how noisy these log lines are. If they only end up being logged in exceptional circumstances we can probably just leave it this way.

Reply:
Sounds like a good plan.